File: src/python/src/leela/driver.py
# -*- coding: utf-8 -*-

import sys
import json
from leela import lql
from leela.exception import LeelaError

def read_lines(fd):
    while (True):
        l = fd.readline()
        if (l == ""):
            break
        yield(json.loads(l)[0].encode("ascii"))

def run():
    session = json.loads(sys.stdin.readline())
    with lql.with_context([e.encode("ascii") for e in session["endpoint"]]) as ctx:
        for stmt in read_lines(sys.stdin):
            with lql.with_cursor(ctx, session["username"], session["secret"], session["timeout"]) as cursor:
                cursor.execute(stmt)
                while (cursor.next()):
                    try:
                        row = cursor.fetch()
                    except LeelaError, e:
                        row = ("fail", e.code, e.message)
                    sys.stdout.write("%s\n" % json.dumps([row]))
                    sys.stdout.flush()
                sys.stdout.write("[null]\n")
                sys.stdout.flush()

if (__name__ == "__main__"):
    run()

File: src/ruby/leela_ruby/spec/bin/driver.rb
#!/usr/bin/env ruby

require "json"
require "bundler"
Bundler.require

require_relative "../../lib/leela_ruby"

def writeln(str)
  $stdout.write str
  $stdout.write "\n"
  $stdout.flush
rescue Errno::EPIPE
  # try_leela may close the pipe without reading it completly
end

def exec_with_connection(with_block, conn)
  until $stdin.eof?
    stmt = JSON.parse($stdin.readline)[0]
    begin
      if with_block
        conn.execute(stmt) {|row| writeln [row].to_json}
      else
        conn.execute(stmt).each {|row| writeln [row].to_json}
      end
    rescue Leela::LeelaError => e
      writeln [[:fail, e.code, e.message]].to_json
    end
    writeln "[null]"
  end
end

def run(with_block)
  session = JSON.parse($stdin.readline)
  if with_block
    Leela::Connection.open(session["endpoint"], session["timeout"], session["username"], session["secret"]) do |conn|
      exec_with_connection with_block, conn
    end
  else
    conn = Leela::Connection.new(session["endpoint"], session["timeout"], session["username"], session["secret"])
    begin
      exec_with_connection with_block, conn
    ensure
      conn.close
    end
  end
end

run (ARGV[0] == "stream")
